Update controllers to use v1 API and make v1 the storage version

wallrj opened the pull request:
Part of #3156 

**Release note**:
```release-note
Updated controllers to use v1 API and make v1 the storage version
```

wallrj commented:
/kind feature

wallrj commented:
/retest

meyskens commented:
Looking good, nice that a dozen lines of bash helps with this!

wallrj commented:
/retest

wallrj commented:
/retest

wallrj commented:
/retest

wallrj commented:
/test pull-cert-manager-e2e-v1-19
/test pull-cert-manager-e2e-v1-17
/test pull-cert-manager-e2e-v1-16
/test pull-cert-manager-e2e-v1-15
/test pull-cert-manager-e2e-v1-14
/test pull-cert-manager-e2e-v1-13
/test pull-cert-manager-e2e-v1-12
/test pull-cert-manager-e2e-v1-11

wallrj commented:
> Looking good, nice that a dozen lines of bash helps with this!

Thanks @meyskens. Might need your help tomorrow fixing e2e test failures.

wallrj commented:
/retest

wallrj commented:
/test pull-cert-manager-e2e-v1-19
/test pull-cert-manager-e2e-v1-17
/test pull-cert-manager-e2e-v1-16
/test pull-cert-manager-e2e-v1-15
/test pull-cert-manager-e2e-v1-14
/test pull-cert-manager-e2e-v1-13
/test pull-cert-manager-e2e-v1-12
/test pull-cert-manager-e2e-v1-11

wallrj commented:
/test pull-cert-manager-e2e-v1-19
/test pull-cert-manager-e2e-v1-17
/test pull-cert-manager-e2e-v1-16
/test pull-cert-manager-e2e-v1-15
/test pull-cert-manager-e2e-v1-14
/test pull-cert-manager-e2e-v1-13
/test pull-cert-manager-e2e-v1-12
/test pull-cert-manager-e2e-v1-11

wallrj commented:
/test pull-cert-manager-e2e-v1-19
/test pull-cert-manager-e2e-v1-17
/test pull-cert-manager-e2e-v1-16
/test pull-cert-manager-e2e-v1-15
/test pull-cert-manager-e2e-v1-14
/test pull-cert-manager-e2e-v1-13
/test pull-cert-manager-e2e-v1-12
/test pull-cert-manager-e2e-v1-11

wallrj commented on pkg/apis/certmanager/v1/generic_issuer.go:
```diff
@@ -0,0 +1,85 @@
+/*
```
I added this back in to v1 because there are bunch of controllers that still use it.
It was removed in response to one of my code review comments  in https://github.com/jetstack/cert-manager/pull/3038/#discussion_r449145480 and at that time @munnerz  wrote:

> This is used by the 'issuers' and 'clusterissuers' controllers so that they can work with both Issuer types and ClusterIssuer (by having them both implement 'GenericIssuer').
> 
> The signatures returned version types (e.g. `GetSpec()` returns a _v1beta1_ IssuerSpec etc).
> 
> It's not actually used right now in this API version, so we can remove it. I agree it shouldn't be in this package, and I think it _should_ be possible to 'refactor this away' with some better written controllers for issuers.
> 
> Will remove from v1beta1, but leaving in v1alpha3 and 2 as I don't want to make a retroactive change in this PR smile

I considered doing that refactoring here, but decided it would cause too much code churn and I'm not really familiar enough with the code to do a good job so close to the v1 release. 

See https://github.com/jetstack/cert-manager/issues/3200

wallrj commented on pkg/apis/certmanager/v1/types.go:
```diff
@@ -184,3 +184,13 @@ const (
 	UsageMicrosoftSGC       KeyUsage = "microsoft sgc"
 	UsageNetscapeSGC        KeyUsage = "netscape sgc"
 )
+
```
This had been removed in v1alpha3 or v1beta1 (can't remember which) but is still used by the controllers.
Perhaps it belongs somewhere internal?

wallrj commented on cmd/controller/app/controller.go:
```diff
@@ -29,7 +29,7 @@ import (
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/kubernetes/scheme"
-	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	clientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
```
My hacky bash script (see next commit) replaces v1alpha2 imports with `v1`, but that sometimes clashes with the corev1 imports. So I'm renaming those.

wallrj commented on hack/change-stored-version.sh:
```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env bash
```
This was something I hacked together to help incrementally update all the v1alpha2 references.
I should probably remove this from the PR once it's been approved.

meyskens replied:
would be nice to have this somewhere for v2 ;)

wallrj replied:
I'll move it to another branch.

wallrj commented on cmd/controller/app/controller.go:
```diff
@@ -23,7 +23,7 @@ import (
 	"sync"
 	"time"
 
```
All the changes in this commit are mechanical. Produced by the script in the previous commit.

wallrj commented on cmd/controller/app/options/BUILD.bazel:
```diff
@@ -7,7 +7,7 @@ go_library(
     visibility = ["//visibility:public"],
```
All mechanical changes in this commit too.

wallrj commented on pkg/controller/certificaterequests/acme/acme_test.go:
```diff
@@ -194,7 +194,7 @@ func TestSign(t *testing.T) {
 				CertManagerObjects: []runtime.Object{baseCR.DeepCopy(), baseIssuer.DeepCopy()},
 				ExpectedEvents: []string{
-					"Normal OrderCreated Created Order resource default-unit-test-ns/test-cr-3921610499",
+					"Normal OrderCreated Created Order resource default-unit-test-ns/test-cr-1733622556",
```
These tests all use resource names derived from a hash of the json representation of the CR.
But because the CR fields have changed, the hash has changed.

wallrj commented on pkg/controller/certificaterequests/acme/acme.go:
```diff
@@ -97,7 +97,7 @@ func (a *ACME) Sign(ctx context.Context, cr *v1.CertificateRequest, issuer v1.Ge
 	if len(csr.Subject.CommonName) > 0 && !util.Contains(csr.DNSNames, csr.Subject.CommonName) {
 		err = fmt.Errorf("%q does not exist in %s", csr.Subject.CommonName, csr.DNSNames)
-		message := "The Request PEM requests a commonName that is not present in the list of dnsNames. If a commonName is set, ACME requires that the value is also present in the list of dnsNames"
+		message := "The CSR PEM requests a commonName that is not present in the list of dnsNames. If a commonName is set, ACME requires that the value is also present in the list of dnsNames"
```
This was changed by the script but this particular change makes the message less clear, so I reverted it.

wallrj commented on pkg/controller/acmeorders/util_test.go:
```diff
@@ -124,7 +124,7 @@ func TestChallengeSpecForAuthorization(t *testing.T) {
 			},
 			expectedChallengeSpec: &cmacme.ChallengeSpec{
-				Type:    "http-01",
+				Type:    cmacme.ACMEChallengeTypeHTTP01,
```
In v1 these types are all upper case strings. I decided to use the constant too, to avoid this problem if the type strings ever change.

wallrj commented on pkg/controller/certificates/internal/test/test.go:
```diff
@@ -79,7 +79,11 @@ func MustCreateCryptoBundle(t *testing.T, crt *cmapi.Certificate, fixedClock *fa
+	crt := originalCert.DeepCopy()
+	if crt.Spec.PrivateKey == nil {
+		crt.Spec.PrivateKey = &cmapi.CertificatePrivateKey{}
+	}
```
What I really want is to create a `SetCertificateDefaults` function for Certificate and call that  everywhere immediately after a Certificate is created or after it has been retrieved from the API server.
Which would avoid having to check for nil everywhere.
I could do it in this PR or I could do it in a follow up.

wallrj commented on deploy/crds/crd-certificaterequests.v1beta1.yaml:
```diff
@@ -49,7 +49,7 @@ spec:
   subresources:
     status: {}
-  version: v1alpha2
+  version: v1
```
This field only exists in the v1beta1 CRD. 
This change is not strictly related to storage version, but without this change, the *filtered* CRD (see next commit) could not be installed because the resulting CRD file had `version: v1alpha2` but a single `versions: ` entry which has a different version.

wallrj commented on pkg/controller/acmechallenges/scheduler/scheduler_test.go:
```diff
@@ -41,7 +41,7 @@ func randomChallenge(rand int) *cmacme.Challenge {
 	return gen.Challenge("test-"+util.RandStringRunes(10),
 		gen.SetChallengeDNSName(util.RandStringRunes(rand)),
-		gen.SetChallengeType("http-01"))
+		gen.SetChallengeType(cmacme.ACMEChallengeTypeHTTP01))
```
I decided to change the signature of SetChallengeType, to avoid this problem in the future.

wallrj commented on devel/setup-e2e-deps.sh:
```diff
@@ -52,4 +54,14 @@ echo "Installing ingress-nginx into the cluster..."
+    echo "ERROR: ${EXIT} setup jobs failed. Check logs above for details."
+fi
+
+exit $EXIT
```
The sample-webhook wasn't getting deployed on K8S < 1.15 because it included v1alpha2 resources which are no longer defined by the legacy CRD yamls.

But this script exited with `0` because bare `wait` *always* exits with 0.

With this change, the script should exit with a non-zero code if any of the background jobs fails, and fail the e2e test early, with a more useful error message.

It could be improved further though.

meyskens replied:
Good catch!

wallrj reviewed (commented):
Self-review and some notes for other reviewers.

JoshVanL commented on cmd/ctl/pkg/create/certificaterequest/certificaterequest.go:
```diff
@@ -38,7 +38,7 @@ import (
 	"github.com/jetstack/cert-manager/cmd/ctl/pkg/util"
 	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
-	cmapiv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmapiv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
```
nit: cmapiv1 -> cmapi

JoshVanL commented on cmd/ctl/pkg/util/util.go:
```diff
@@ -22,16 +22,16 @@ import (
 
 	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
-	cmapiv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmapiv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
```
nit: cmapiv1 -> cmapi

JoshVanL commented on pkg/internal/apis/certmanager/validation/certificate_test.go:
```diff
@@ -25,7 +25,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/validation/field"
 
-	cmapiv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmapiv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
```
nit: cmapiv1 -> cmapi

wallrj replied:
Done. Although it seems a bit fishy that we're importing and using the external API version in the internal version.

JoshVanL commented on cmd/ctl/pkg/convert/convert.go:
```diff
@@ -34,7 +34,7 @@ import (
 	"k8s.io/kubectl/pkg/util/templates"
 
-	cmapiv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmapiv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
```
nit: cmapiv1 -> cmapi

JoshVanL commented on cmd/ctl/pkg/status/certificate/types.go:
```diff
@@ -30,7 +30,7 @@ import (
 	"github.com/jetstack/cert-manager/cmd/ctl/pkg/status/util"
 	cmacme "github.com/jetstack/cert-manager/pkg/apis/acme/v1beta1"
-	cmapiv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmapiv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
```
nit: cmapiv1 -> cmapi

JoshVanL commented on pkg/internal/apis/certmanager/validation/certificate.go:
```diff
@@ -25,7 +25,7 @@ import (
 
 	"github.com/jetstack/cert-manager/pkg/api/util"
-	cmapiv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmapiv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
```
nit: cmapiv1 -> cmapi

JoshVanL commented on pkg/internal/apis/certmanager/validation/certificate_test.go:
```diff
@@ -25,7 +25,7 @@ import (
 
-	cmapiv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmapiv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
 	cmapi "github.com/jetstack/cert-manager/pkg/internal/apis/certmanager"
```
cmapi -> internalcmapi

JoshVanL commented on pkg/internal/apis/certmanager/validation/certificate.go:
```diff
@@ -25,7 +25,7 @@ import (
 	"github.com/jetstack/cert-manager/pkg/api/util"
-	cmapiv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmapiv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
 	cmapi "github.com/jetstack/cert-manager/pkg/internal/apis/certmanager"
```
nit: cmapi -> internalcmapi

JoshVanL reviewed (changes requested):
Just a few nits :)

meyskens commented on hack/change-stored-version.patch:
```diff
@@ -0,0 +1,112 @@
++					IssuerRef:  issuerRef,
+ 				},
+ 			}
+ 			By("Creating a Certificate")
```
Should we keep this patch file?

wallrj replied:
I'll remove it along with the rest of the change-stored-version script.

meyskens commented on pkg/controller/acmeorders/sync.go:
```diff
@@ -424,16 +424,16 @@ func (c *controller) finalizeOrder(ctx context.Context, cl acmecl.Interface, o *
 
 	// Due to a bug in the initial release of this controller, we previously
-	// only supported DER encoded CSRs and not PEM encoded as they are intended
+	// only supported DER encoded Requests and not PEM encoded as they are intended
```
I think the comment can keep CSR ;)

meyskens commented on pkg/controller/certificaterequests/acme/acme.go:
```diff
@@ -74,27 +74,27 @@ func NewACME(ctx *controllerpkg.Context) *ACME {
+	// If we can't decode the Request PEM we have to hard fail
+	csr, err := pki.DecodeX509CertificateRequestBytes(cr.Spec.Request)
 	if err != nil {
-		message := "Failed to decode CSR in spec"
```
can we use CSR here also?

wallrj commented:
Thanks for your reviews @meyskens and @JoshVanL 

I've addressed your points and removed the bash script and all the associated filed to #3201 

There are still a bunch of references to v1alpha2 in the repo, some of which should probably be updated, but I propose we do that in another PR:

```

cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:apiVersion: cert-manager.io/v1alpha2
cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:apiVersion: cert-manager.io/v1alpha2
cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:apiVersion: cert-manager.io/v1alpha2
cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:apiVersion: cert-manager.io/v1alpha2
deploy/charts/cert-manager/README.template.md:> This version does not have API version conversion enabled and only supports `cert-manager.io/v1alpha2` API resources.
design/release-notes/release-0.11/draft-release-notes.md:| Certificate        | `certmanager.k8s.io/v1alpha1` | `cert-manager.io/v1alpha2`      |
design/release-notes/release-0.11/draft-release-notes.md:| Issuer             | `certmanager.k8s.io/v1alpha1` | `cert-manager.io/v1alpha2`      |
design/release-notes/release-0.11/draft-release-notes.md:| ClusterIssuer      | `certmanager.k8s.io/v1alpha1` | `cert-manager.io/v1alpha2`      |
design/release-notes/release-0.11/draft-release-notes.md:| CertificateRequest | `certmanager.k8s.io/v1alpha1` | `cert-manager.io/v1alpha2`      |
design/release-notes/release-0.11/draft-release-notes.md:| Order              | `certmanager.k8s.io/v1alpha1` | `acme.cert-manager.io/v1alpha2` |
design/release-notes/release-0.11/draft-release-notes.md:| Challenge          | `certmanager.k8s.io/v1alpha1` | `acme.cert-manager.io/v1alpha2` |
test/integration/ctl/ctl_convert_test.go:       targetv1alpha2 = "cert-manager.io/v1alpha2"
test/integration/ctl/ctl_convert_test.go:apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/ctl_convert_test.go:- apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/ctl_convert_test.go:- apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/ctl_convert_test.go:- apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/ctl_convert_test.go:- apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/testdata/convert_resource1.yaml:apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/testdata/convert_resource2.yaml:apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/testdata/convert_resource2.yaml:apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/testdata/convert_resource2.yaml:apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/testdata/convert_resource3.yaml:apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/testdata/convert_resource_with_organization_v1alpha2.yaml:apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/testdata/create_cr_cert_with_ns1.yaml:apiVersion: cert-manager.io/v1alpha2
test/integration/ctl/testdata/create_cr_issuer.yaml:apiVersion: cert-manager.io/v1alpha2
```

meyskens commented:
I think only the following should change:
```
cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:apiVersion: cert-manager.io/v1alpha2
cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:apiVersion: cert-manager.io/v1alpha2
cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:apiVersion: cert-manager.io/v1alpha2
cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:apiVersion: cert-manager.io/v1alpha2
deploy/charts/cert-manager/README.template.md:> This version does not have API version conversion enabled and only supports `cert-manager.io/v1alpha2` API resources.
```

the others are either tests which use conversion so that is fine and the should still be tested

meyskens commented:
/approve
/lgtm
